app/lib/animation.ts

```diff
@@ -0,0 +1,123 @@
+import { clamp } from "./math";
+import { TimeDelta } from "./time";
+
+export class AnimationLoop {
+  private dt = new TimeDelta();
+
+  constructor(private callback: (dt: number) => void, private enabled_ = false) {
+    if (enabled_) {
+      this.start();
+    }
+  }
+
+  private loop() {
+    if (!this.enabled_ || this.isNextFrameScheduled) return;
+
+    this.callback(this.dt.get());
+    this.scheduleNextFrame();
+  }
+
+  private isNextFrameScheduled = false;
+  private scheduleNextFrame() {
+    if (this.isNextFrameScheduled) return;
+    this.isNextFrameScheduled = true;
+    requestAnimationFrame(() => {
+      this.isNextFrameScheduled = false;
+      this.loop();
+    });
+  }
+
+  start() {
+    this.enabled_ = true;
+    this.loop();
+  }
+
+  stop() {
+    this.enabled_ = false;
+    this.dt.reset();
+  }
+}
+
+export class AnimatedValue {
+  private target_: number;
```

Comment:
if we're making things private, do we need the underscores?

Reply:
The underscores are shorthand for "internal" -- they differentiate between the public getter/setter and the private internal value.